Add support for cluster-scoped AdminNetworkPolicy resource (#2522)
kubernetes/enhancements: keps/sig-network/2091-cluster-network-policy/README.md

security rules for the cluster, i.e. a developer CANNOT override these rules
by creating NetworkPolicies that applies to the same workloads as the
ClusterNetworkPolicy does.

Unlike the NetworkPolicy resource in which each rule represents an allowed
traffic, ClusterNetworkPolicy will enable administrators to set `Empower`,
`Deny` or `Allow` as the action of each rule. ClusterNetworkPolicy rules should
be read as-is, i.e. there will not be any implicit isolation effects for the Pods
selected by the ClusterNetworkPolicy, as opposed to what NetworkPolicy rules imply.

In terms of precedence, the aggregated `Empower` rules (all ClusterNetworkPolicy
rules with action `Empower` in the cluster combined) should be evaluated before
aggregated ClusterNetworkPolicy `Deny` rules, followed by aggregated ClusterNetworkPolicy
`Allow` rules, followed by NetworkPolicy rules in all Namespaces. As such, the
`Empower` rules have the highest precedence, which shall only be used to provide
exceptions to deny rules. The `Empower` rules do not guarantee that the traffic
will not be dropped: it simply denotes that the packets matching those rules can bypass the

> ClusterNetworkPolicy `Deny` rule evaluation.

Reviewer: We were discussing this on the sig-network-policy meeting yesterday, and I want to clarify the meaning of `Empower` as it stands a bit more here:
There has been two schools of thought on how `Empower` should be interpreted: 1. `Empower` provides exception to Deny rules of ClusterNetworkPolicy  2. `Empower` delegates policy control of specific traffic to the developers. While 1. is always true,  the delegation happens when there are no ClusterNetworkPolicy Allow rules explicitly dictating that traffic must be allowed (which must be a subset of what's Empowered, in the presence of some sort of Deny-all rule). Consider the following example (assuming each tenant owns single namespace):
 ```
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        type: tenant
  ingress:
    - action: Empower
      from:
        - namespaceSelector:
            matchLabels:
               app: prometheus
        - namespaces:
             scope: Self
    - action: Deny
      from:
        - namespaceSelector: {}
    - action: Allow
      from:
        - namespaceSelector:
             matchLabels:
               app: prometheus   
```
There are 3 different kind of traffic affected: 
1. Prometheus namespace -> tenant namespace: Always allowed, as it is under `Empower` AND `Allow`. This might be a requirement for compliance purposes for example.

2. tenant A -> tenant A, tenant B -> tenant B etc: Delegated to developers of each tenant namespace. Those traffic are not Denied because of `Empower`, but not explicitly allowed either. Namespace owners can use k8s NetworkPolicies to further regulate intra-namespace traffic.

3. tenant B -> tenant A, tenant A -> tenant B etc: Dropped.

Hope it will clarify things a bit more, and feedbacks are also welcomed.

> ClusterNetworkPolicy `Deny` rule evaluation.

Author: resolving as this is regarding an older proposal we had

> # KEP-2091: Add support for ClusterNetworkPolicy resources

Reviewer 2: Note to add possible name change to `admin network policy`.

Reviewer 2: Done

[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Introduce new set of APIs to express an administrator's intent in securing
their K8s cluster. This doc proposes the ClusterNetworkPolicy API to complement
the developer focused NetworkPolicy API in Kubernetes.

## Motivation

Kubernetes provides NetworkPolicy resources to control traffic within a
cluster. NetworkPolicies focus on expressing a developers intent to secure
their applications. Thus, in order to satisfy the needs of a security admin,
we propose to introduce a new API that captures the administrators intent.

### Goals

The goals for this KEP are to satisfy the following key user stories:
1. As a cluster administrator, I want to enforce irrevocable guardrails that
   all workloads must adhere to in order to guarantee the safety of my clusters. 
   In particular I want to enforce certain network level access controls that are 
cluster scoped and cannot be overridden or bypassed by namespace scoped network policies. 
2. As a cluster administrator, I want to have the option to enforce network level access controls 

> that facilitate network multi-tenancy and strict network level isolation between multiple 

Reviewer 2: nit: indent

kubernetes cluster to another.

####  Multiple selectors per policy instance

With the current Kubernetes network policy v1 api, an instance of a policy can be applied
to a single set of pods defined using a single selector instance (which can be one of 3 types).
However with the proposed CNP specification, it shall be possible to have multiple instances
of selectors within a single CNP policy instance and each such selector can be configured
with its own set of peer endpoints and policy actions. This enables greater flexibility
of configuration, and improves ease of use by not requiring cluster admins to have to come up
with unique priority values for each instnce of rule while retaining explicit rule ordering
as described earlier since multiple rules wihin a CNP instance are prioritized by their listing 
order within the CNP manifest. A fully specified new rule (including selector, peer and action 
type definitions) can hence be unambiguously inserted into an existing configuration of CNPs and 
rules without requiring any reshuffling of priority values and while retaining unambiguous 
relative priority ordering of all configured rules and CNP instances. 

> #### Default rules

Reviewer 2: I think we go ahead and convert this to Yang's idea of a "default" enum

	// empty or missing, this rule matches all ports (traffic not restricted by port).
	// If this field is present and contains at least one item, then this rule allows/denies
	// traffic only if the traffic matches at least one port in the list.
	// +optional
	Ports        []networkingv1.NetworkPolicyPort
	// List of sources/dest which should be able to access the pods selected for this rule.
	// Items in this list are combined using a logical OR operation. If this field is
	// empty or missing, this rule matches all sources/dest (traffic not restricted by
	// source/dest). If this field is present and contains at least one item, this rule
	// allows/denies traffic only if the traffic matches at least one item in the from/to list.
	// +optional
	From/To      []ClusterNetworkPolicyPeer
	// Name is an identifier for this rule. Must be unique within an instance of CNP. Can be
	// same across policy instances.
	Name         string
}

type ClusterNetworkPolicyType string

// Add future selector types here to ease introduction of future selectors to ensure api fails 
// closed cleanly if a CNI doesnt support specific types of selectors

> const (

Reviewer 2: @Dyanngg Please Update :)

3. Backwards compatibility with existing Kubernetes Network Policy API

### Non-Goals

Our mission is to solve the most common use cases that cluster admins have. That is, we don't want to solve for every possible policy permutation a user can think of. Instead, we want to design an API that addresses 90-95% use cases while keeping the mental model easy to understand and use.

Additionally, this proposal is squarely focused on solving the needs of the Cluster Administrator. It is not intended to solve:
1. Logging / error reporting for Network Policy
2. Kubernetes Node policies
3. New policy selectors (services, service accounts, etc.)

## Proposal

In order to achieve the two primary broad use cases for a cluster admin to secure K8s clusters,
we propose to introduce the following two new resources under `networking.k8s.io` API group:
- ClusterNetworkPolicy resource
- DefaultNetworkPolicy resource

### ClusterNetworkPolicy resource

A ClusterNetworkPolicy resource will help the administrators set strict security rules for the cluster,
i.e. a developer CANNOT override these rules by creating NetworkPolicies that applies to the same workloads
as the ClusterNetworkPolicy does.

> Unlike the NetworkPolicy resource in which each rule represents a whitelisted traffic, ClusterNetworkPolicy

Reviewer 3: would change the term whitelisted to allow listed here (or maybe some other term)

Author: thanks.. done

- [ ] (R) Design details are appropriately documented
- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
- [ ] (R) Graduation criteria is in place
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Introduce new set of APIs to express an administrator's intent in securing
their K8s cluster. This doc proposes the ClusterNetworkPolicy API to complement
the developer focused NetworkPolicy API in Kubernetes.

> (TODO) In the multi-cluster context, we are thinking to name the extended NetworkPolicy

Reviewer 2: Ack thanks

to MultiClusterNetworkPolicy. To avoid the extension of this API resource to be named
MultiCluster-ClusterNetworkPolicy, maybe we should name this resource differently
to begin with? AdminNetworkPolicy is a popular alternative.

## Motivation

Kubernetes provides NetworkPolicy resources to control traffic within a
cluster. NetworkPolicies focus on expressing a developers intent to secure
their applications. Thus, in order to satisfy the needs of a security admin,

> we propose to introduce a new API that captures the administrators intent.

Reviewer 2: Agreed, I'll try and add a bit more here

### Goals

> The goals for this KEP are to satisfy the following key user stories:

Reviewer 2: I tried to add examples, however the sort of it is that 1 and 2 are closely intertwined however 1 is related to expressing simple allow-deny semantics at a cluster level, while 2 is related to building in cluster tenants.

1. As a cluster administrator, I want to enforce irrevocable guardrails that
   all workloads must adhere to in order to guarantee the safety of my clusters.
   In particular I want to enforce certain network level access controls that are
cluster scoped and cannot be overridden or bypassed by namespace scoped network policies.
2. As a cluster administrator, I want to have the option to enforce network level access controls
   that facilitate network multi-tenancy and strict network level isolation between multiple
   teams and tenants sharing a cluster via use of namespaces or groupings of namespaces per tenant.
3. As a cluster administrator, I want to optionally also deploy an additional default set of policies to all
   workloads that may be overridden by the developers if needed.

There are several unique properties that we need to add in order accomplish the
user stories above.
1. Deny rules and, therefore, hierarchical enforcement of policy
2. Semantics for a cluster-scoped policy object that may include
   namespaces/workloads that have not been created yet.

> 3. Backwards compatibility with existing Kubernetes Network Policy API

Reviewer 2: +1

### Non-Goals

Our mission is to solve the most common use cases that cluster admins have.
That is, we don't want to solve for every possible policy permutation a user
can think of.  Instead, we want to design an API that addresses 90-95% use cases
while keeping the mental model easy to understand and use.
The focus of this KEP is on cluster scoped controls for east-west traffic within
a cluster. Cluster scoped controls for North-South traffic may be addressed via

> future versions of the api resources introduced in this or other future KEPs.

Reviewer 4: I don't know if it goes here or not, but somewhere it should be mentioned explicitly that CNPs never affect north-south traffic, and thus also don't override or bypass NetworkPolicies with ipBlock rules that select external traffic

Reviewer: Will add this

Reviewer 4: It doesn't need to explicitly ignore it; CNP rules can only match traffic which has a pod as both its source and its destination, so any N-S traffic will necessarily not match any CNPs and thus not be acted upon.

Additionally, this proposal is squarely focused on solving the needs of the
Cluster Administrator. It is not intended to solve:
1. Error reporting for Network Policy
2. Kubernetes Node policies
3. New policy selectors (services, service accounts, etc.)
4. Cluster external traffic access control.

## Proposal

In order to achieve the two primary broad use cases for a cluster admin to
secure K8s clusters, we propose to introduce the following resource
under `netpol.networking.k8s.io` API group:

> - ClusterNetworkPolicy

Reviewer 2: There was some discussion on this, but I think the WG is pretty flexible as it pertains to naming.  CC @abhiraut @srampal @Dyanngg

> - ClusterNetworkPolicy

Author: this was a placeholder .. yes.. `policy` sounds good... there were other suggestions too.. I think `security` was one?

> - ClusterNetworkPolicy

Reviewer 4: So is the next feature to come along going to run into this same problem? (Having to come up with a new group name because it can't reuse any existing group name?) Or is this something that now that we know the problem exists, we can do things differently to avoid killing off `v1beta1` in the new group?

Anyway, if this group is going to end up being for CNP only and nothing else, then it should have a more "specific" name (like `policy`) rather than using up a more "generic" name (like `security`) that might be better used for some other feature later.

> - ClusterNetworkPolicy

Reviewer 2: I thing we should try to keep v1beta1 alive for the `policy` group, it doesn't make sense to have a new group for every new resource that comes out of the sig-network-policy API subgroup

> - ClusterNetworkPolicy

Reviewer 4: I don't remember the details... this was discussed a while back (either in the context of CNP or Ingress v2), and someone said that because of the way past API deprecations had been handled, that `networking.k8s.io/v1beta1` cannot be brought back.

### ClusterNetworkPolicy resource

A ClusterNetworkPolicy (CNP) resource will help the administrators set strict
security rules for the cluster, i.e. a developer CANNOT override these rules
by creating NetworkPolicies that applies to the same workloads as the

> ClusterNetworkPolicy does.

Reviewer 2: Agreed, tried to clear this up

ClusterNetworkPolicy `Deny` rules are useful for administrators to explicitly
block traffic from malicious clients, or workloads that poses security risks.
Those traffic restrictions can only be lifted once the `Deny` rules are deleted
or modified by the admin.

On the other hand, the `Allow` rules can be used to call out traffic in the cluster
that needs to be allowed for certain components to work as expected (egress to
CoreDNS for example). Those traffic should not be blocked when developers apply
NetworkPolicy to their Namespaces which isolates the workloads.

ClusterNetworkPolicy `Pass` rules allows an admin to delegate security posture for
certain traffic to the Namespace owners. For example, allowing intra-tenant traffic
can be delegated to tenant admins explicitly by the cluster admin with the use of `pass`
rules.

#### Priority

The policy instances will be ordered based on the numeric priority assigned to the
CNP. A lower number corresponds to a higher order. For example, CNP with `priority` set
to "0" will be considered as the highest priority policy, while a large number, i.e.
999, will be considered as low priority. Any numeric-valued priority, however, will have

> higher precedence over the namespaced NetworkPolicy instances in the cluster.

Reviewer 2: I tend to lean towards keeping this simple and having 0 be the lowest 999 be the highest.... The main goal with flipping this was so that it would be easier to write  an "absolute" CNP... I.e I know if I have a priority 0 CNP to deny traffic to some pods that it will never be overridden

> higher precedence over the namespaced NetworkPolicy instances in the cluster.

Reviewer 5: For what it is worth, in Calico we call this field `order`, to indicate that policies are applied in order - i.e. 0 before 1, etc. I'm personally in favor of using that model here (independent of the name) - 0 is the first applied, and larger numbers have a lower priority.

What's the reason for choosing 999 as the highest? Seems arbitrary, and could be limiting for larger / more complex operators who might want to further build an ordering scheme within the allowed range. I'd suggest just allowing a uint32 here.

> higher precedence over the namespaced NetworkPolicy instances in the cluster.

Reviewer 2: I like the idea of using `order` as well! @caseydavenport The 999 limit was to ensure portability of CNP's between clusters, since some CNI's may only support a specific number of policies (for example OVN ACLs only support prioities between between 0 and 32767) however we assumed all CNIs should be able to support 1000 unique priorities.  Also we allow up to 100 rules within each policy so we figured that there should be plenty of space for admins to express their intent 

Lastly we believe there shouldn't be anywhere close to that many CNPs deployed in a cluster because typically there is only a single or small group of admins who would be writing/ managing those.

> higher precedence over the namespaced NetworkPolicy instances in the cluster.

Author: +1 to Dan's suggestion on usage of words (in his edits) .. 

@caseydavenport @astoycos  I personally do not prefer 999(or any other arbitrary number) as highest. Would continue to suggest we keep 0 as the policy which takes highest precedence and so on..

> higher precedence over the namespaced NetworkPolicy instances in the cluster.

Reviewer 6: I lean towards 0 being the lowest importance, too.  I like "order" as a  clarifying name.  It also lets us use 0 as a  special value (baseline).

Additionally, a special keyword `Baseline` can be used in the priority field to indicate
that the rules in that policy instance shall be created at a priority lower than the Namespaced

> NetworkPolicies. Admission controllers will need to validate that only this string value or

Reviewer 2: No I think this was a mistake, CRD validation should be fine here

> NetworkPolicies. Admission controllers will need to validate that only this string value or

Reviewer 6: As discussed on sig-net call - let's NOT do IntOrString.  please

The precedence of the rules within a CNP will be determined by the order in which the rule
is written. Thus, a rule that appears at the top of the ingress/egress rules would take the
highest precedence.

Conflict resolution: Two policies are considered to be conflicting if they are assigned the
same `priority`. In order to avoid such conflicts, we propose to include validation for
CNP resources, such that no two CNP instances can be created with the same `priority`.
(TODO: Is such admission control viable?)

For conformance and portability purposes, it will be required that implementations
support up to 1000 CNP priorities (0 through 999) per cluster and up to 100 rules
within a CNP instance.  An implementation can optionally choose to support larger numbers
of priority values and rules, however if an implementation supports a
smaller number of rules and instances than these recommended ranges, it will be considered
non-conformant.  This ensures that policy manifests written using priorities 0 through
999 for instance are portable across clusters and implementations without need to re-write
policy manifests or re-adjust priority allocations when moving applications from 1

> kubernetes cluster to another.

Reviewer 4: I don't think it makes sense to say that a larger range exists but only a subset is portable. If I want to create a "minimum-priority" policy, then to be portable I would have to make it be `Priority: 999`, but at that point it's no longer possible to use any `Priority` value larger than that.

You can say "implementations must support at least 1000 distinct `Priority` values" without also saying that those values have to be 0-999. Either that or we should just say that values outside the range 0-999 are not valid.

Reviewer 2: Done

####  Multiple selectors per policy instance

With the current Kubernetes network policy v1 api, an instance of a policy can be applied

> to a single set of pods defined using a single selector instance (which can be one of 3 types).

Reviewer 2: I think the 3 types here was 1. PodSelector 2. NamespaceSelector 3. PodSelector+NamespaceSelector

> to a single set of pods defined using a single selector instance (which can be one of 3 types).

Reviewer 4: Those are three types of ingress/egress rules, but (AFAICT) this is talking about `spec.podSelector`, which is, as the name suggests, a pod selector.

rules without requiring any reshuffling of priority values and while retaining unambiguous
relative priority ordering of all configured rules and CNP instances.

#### Rule identifiers

In order to collect traffic statistics corresponding to
a rule, it is necessary to identify the rule which allows/denies that traffic.
This helps administrators figure the impact of the rules written in a
cluster-scoped policy resource. Thus, the ability to uniquely identify a rule
within a cluster-scoped policy resource becomes very important.
We propose to introduce a new string field, called `name`, per `ClusterNetworkPolicy`
ingress/egress rule. The `name` of a rule must be unique withtin the same CNP. However,
two CNP instances could have rules with same names.

### User Stories

Note: This KEP will focus on East-West traffic, cluster internal, user stories and
not address North-South traffic, cluster external, use cases, which will be
solved in a follow-up proposal.

#### Story 1: Deny traffic at a cluster level

As a cluster admin, I want to apply non-overridable deny rules
to certain pod(s) and(or) Namespace(s) that isolate the selected
resources from all other cluster internal traffic.

> ![Alt text](explicit_isolation.png?raw=true "Explicit Deny")

Reviewer 2: This all makes sense....I will work to make these diagrams/sample yamls a bit cleaner/ easier to understand

delegated traffic the traffic will be allowed.

![Alt text](delegation.png?raw=true "Delegate")

#### Story 4: Create and Isolate multiple tenants in a cluster

As a cluster admin, I want to build tenants (modeled as Namespace(s))
in my cluster that are isolated from each other by default. Tenancy may be
modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a
single tenant may own more than 1 Namespace.

![Alt text](tenants.png?raw=true "Tenants")

#### Story 5: Cluster Wide Default Guardrails

As a cluster admin I want to change the default security model for my cluster.

In order to follow best practices, an admin may want to spawn a cluster with default
rules blocking all intra cluster traffic. Only traffic that is essential to the cluster
will be opened up with stricter cluster network policies. While Namespace owners will be
forced to use NetworkPolicies to explicitly allow only known traffic. This follows
a explicit allow model which is familiar to many security administrators, and similar
to how [kubernetes suggests network policy be used](https://kubernetes.io/docs/concepts/services-networking/network-policies/#default-policies).

> ![Alt text](default_rules.png?raw=true "Default Rules")

Reviewer 2: Agreed, went ahead and added another

|--------------------------|------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------|
| Target persona           | Cluster administrator or equivalent                                                                              | Developers within Namespaces                                                       |
| Scope                    | Cluster                                                                                                          | Namespaced                                                                         |
| Drop traffic             | Supported with a `Deny` rule action                                                                              | Supported via implicit isolation of target Pods                                    |
| Skip enforcement         | Supported with an `Pass` rule action                                                                             | Not needed                                                                         |

> | Allow traffic            | Supported with an `Allow` rule action for ClusterNetworkPolicy                                                   | Default action for all rules is to allow                                           |

Reviewer 2: Done

| Implicit isolation       | No implicit isolation                                                                                            | All rules have an implicit isolation of target Pods                                |
| Rule precedence          | Depends on the order in which they appear within a CNP                                                           | Rules are additive                                                                 |
| Policy precedence        | Depends on `priority` field among CNPs. Enforced before K8s NetworkPolicies if numeric priority value            | Enforced after numeric-priority ClusterNetworkPolicies, before basline-priority ClusterNetworkPolicy |
| Matching pod selection   | Multiple "Applied-to" selector fields supported per CNP instance                                                 | Single "Applied-to" selector per policy instance                                   |
| Rule identifiers         | Name per rule in string format. Unique within a CNP.                                                             | Not supported                                                                      |

> | Rule identifiers         | Name per rule in string format. Unique within a CNP.                                                             | Not supported                                                                      |

Reviewer 2: Done

> | Cluster external traffic | Not supported                                                                                                    | Supported via IPBlock                                                              |

Reviewer 2: Done

| Namespace selectors      | Supports advanced selection of Namespaces with the use of `namespaces`and label based `namespaceSelector` fields | Supports label based Namespace selection with the use of `namespaceSelector` field |

> 

Reviewer 2: Done

did not take effect, even if they know to look for ClusterNetworkPolicy rules
that can potentially override these policies:
To understand why traffic between a pair of Pods is allowed/denied, a list of
NetworkPolicy resources in both Pods' Namespace used to be sufficient
(considering no other CRDs in the cluster tries to alter traffic behavior).
The same Pods, on the other hand, can appear as an AppliedTo, or an ingress/egress
peer in any ClusterNetworkPolicy. This makes looking up policies that affect a
particular Pod more challenging than when there's only NetworkPolicy resources.

In case where a ClusterNetworkPolicy `Pass` rule,
ClusterNetworkPolicy `Deny` rule, ClusterNetworkPolicy `Allow` rule,
NetworkPolicy rule and default rule applies to an overlapping
set of Pods, users will need to refer to the priority associated with the
rule, to determine which rule would take effect. Figuring out how stacked policies
affect traffic between workloads might not be very straightfoward.

To mitigate this risk and improve UX, a tool which does a reverse look up and lists rules
and policies that impact a given Pod and prints out relative precedence of those rules

> can be quite useful. The [cyclonus](https://github.com/mattfenwick/cyclonus)

Reviewer 2: I think this needs to be done, but I don't think it should be the CNI, this process should stay in tree and not be implementation specific.

> can be quite useful. The [cyclonus](https://github.com/mattfenwick/cyclonus)

Reviewer 2: Admin -> Needs to know what my CNP is going to break and has the ability to see that 

Developer -> Needs to know if my NPs are going to be overridden/ interrupted by any CNPS

> can be quite useful. The [cyclonus](https://github.com/mattfenwick/cyclonus)

Reviewer 2: Need to add more detail to this section

> can be quite useful. The [cyclonus](https://github.com/mattfenwick/cyclonus)

Reviewer 4: So... I guess even without special privileges, we can always answer the question "Is this connection allowed or denied according to _my_ NetworkPolicies?".

Like, if I'm an admin of Namespace A, and I want my pod "foo" to be able to connect to service "bar" in Namespace B, then I ask this tool, and it would say "according to the NPs in Namespace A, this connection is allowed, but it might be blocked by an NP in Namespace B, or a CNP". So then, if the connection _doesn't_ work, at least you know that it's not your fault.

So then you could go to a Namespace B admin and have them run the same check, and the tool would say "according to the NPs in Namespace B, this connection is allowed, but it might be blocked by an NP in Namespace A, or a CNP". And so then at that point, you know it must be blocked by a CNP.

(And a cluster-administrator could run the tool, and it would say "this connection is allowed by the NPs in Namespace A and the NPs in Namespace B but is blocked by CNP X".)

- **Audit Logging**: Very often cluster administrators want to log every connection
  that is either denied or allowed by a firewall rule and send the details to
  an IDS or any custom tool for further processing of that information.
  With the introduction of `deny` rules, it may make sense to incorporate the
  cluster-scoped policy resources with a new field, say `auditPolicy`, to
  determine whether a connection matching a particular rule/policy must be
  logged or not.

## Design Details

### ClusterNetworkPolicy API Design
The following new `ClusterNetworkPolicy` API will be added to the `netpol.networking.k8s.io` API group.

**Note**: Much of the behavior of certain fields proposed below is intentionally aligned with K8s NetworkPolicy
resource, wherever possible. For eg. the behavior of empty or missing fields matches the behavior specified in
the NetworkPolicySpec.

```golang
type ClusterNetworkPolicy struct {
	metav1.TypeMeta
	metav1.ObjectMeta
	// Specification of the desired behavior of ClusterNetworkPolicy.
	Spec ClusterNetworkPolicySpec
}

type ClusterNetworkPolicySpec struct {
	Priority     IntOrString

> 	// No implicit isolation of AppliedTo Pods/Namespaces.

Reviewer 2: Fixed

	// Required field.
	// List of multiple combinations of Selector, Peers and Actions
	PolicyRules  []ClusterNetworkPolicyRule
}

type ClusterNetworkPolicyRule struct {
	AppliedTo    AppliedTo
	Ingress      []ClusterNetworkPolicyIngressRule
	Egress       []ClusterNetworkPolicyEgressRule
}

type ClusterNetworkPolicyIngress/EgressRule struct {
	// Action specifies whether this rule must pass, allow traffic or deny traffic.
	// Required field for any rule.
	Action       RuleAction
	// List of ports for incoming/outgoing traffic.
	// Each item in this list is combined using a logical OR. If this field is
	// empty or missing, this rule matches all ports (traffic not restricted by port).

> 	// If this field is present and contains at least one item, then this rule allows/denies

Reviewer 2: +1

	// traffic only if the traffic matches at least one port in the list.
	// +optional
	Ports        []networkingv1.NetworkPolicyPort
	// List of sources/dest which should be able to access the pods selected for this rule.
	// Items in this list are combined using a logical OR operation. If this field is
	// empty or missing, this rule matches all sources/dest (traffic not restricted by
	// source/dest). If this field is present and contains at least one item, this rule
	// allows/denies traffic only if the traffic matches at least one item in the from/to list.
	// +optional
	From/To      []ClusterNetworkPolicyPeer
	// Name is an identifier for this rule. Must be unique within an instance of CNP. Can be
	// same across policy instances.
	Name         string
}

// WorkloadSelector selects workloads based on the label Selector.
// Right now the only selector type supported is PodSelector. More selector types may be

> // added in the future to support other workload selection mechanism, e.g. serviceAccountSelector.

Reviewer 2: ```
type NamespaceAndServiceSelector struct { 
         NamespaceSelector *metav1.LabelSelector
         ServiceSelector        *metav1.LabelSelector
}
```

// Must be accompanied by a NamespaceSelector to specify the scope of the workload selection.
type WorkloadSelector struct {
	PodSelector         *metav1.LabelSelector
}
```

**Alternatives considered for the WorkloadSelector struct:**  
```golang
type WorkloadSelector struct {
	selectorType        WorkloadSelectorType
	selector            metav1.LabelSelector
}

type WorkloadSelectorType string

// Add future selector types here to ease introduction of future selectors (e.g. serviceAccountSelector)
// and ensure api fails closed cleanly if a CNI doesn't support specific types of selectors.
const (
	PodSelector WorkloadSelectorType  = "PodSelector"
)
```
This alternative WorkloadSelector definition is more explicit but more verbose as well.
At this point, it is expected that any future selectors will be mutually exclusive to the
PodSelector in a single WorkloadSelector. Both approaches should be future extensible
without breaking the premise of fail-closed API design.

(TODO) Does there seem to be a valid usecase where two or more types of selectors will
need to co-exist in a same workloadSelector? i.e. select workloads in these Namespaces,

> where it matches certain Pod labels, AND matches serviceAccount selection criteria?

Reviewer 4: Your user stories should provide the answer to this question. Right now they don't provide any use case for either podSelector or serviceAccountSelector.

```golang
type ClusterNetworkPolicyPeer struct {
	WorkloadSelector    *WorkloadSelector
	// One of NamespaceSelector or Namespaces is required.
	// In the same ClusterNetworkPolicyPeer, NamespaceSelector and Namespaces fields are mutually
	// exclusive.
	NamespaceSelector   *metav1.LabelSelector
	Namespaces          *Namespaces
}

const (
	// RuleActionPass enables admins to provide exceptions to ClusterNetworkPolicies and delegate this rule to
	// K8s NetworkPolicies.
	RuleActionPass      RuleAction = "Pass"
	// RuleActionDeny enables admins to deny specific traffic.
	RuleActionDeny      RuleAction = "Deny"
	// RuleActionAllow enables admins to specifically allow certain traffic.
	RuleActionAllow     RuleAction = "Allow"
)
```

For the ClusterNetworkPolicy ingress/egress rule, the `Action` field dictates whether
traffic should be allowed or denied from/to the ClusterNetworkPolicyPeer. This will be a required field.

The following structs will be added to the `netpol.networking.k8s.io` API group:

```golang
type AppliedTo struct {
	// required if a PodSelector is specified

> 	NamespaceSelector   *metav1.LabelSelector

Reviewer 4: What are the semantics of this field? What are the semantics of leaving it `nil`? (Are they actually different from the semantics of having it set-but-empty?)

	// optional
	WorkloadSelector    *WorkloadSelector
}

// Namespaces define a way to select Namespaces in the cluster.
type Namespaces struct {
	Scope       NamespaceMatchType
	// Labels are set only when scope is "SameLabels".
	Labels      []string
	// Selector is only set when scope is "Selector".
	// Namespaces.Selector has the same effect as NamespaceSelector.

> 	Selector    *metav1.LabelSelector

Reviewer 4: Why have `NamespaceSelector` at all then? Just get rid of it and leave the more expressive `Namespaces` field only.

("Because NetworkPolicy has a NamespaceSelector field" is not a good answer.)

}

// NamespaceMatchType describes Namespace matching strategy.
type NamespaceMatchType string

// This list can/might get expanded in the future (i.e. NotSelf etc.)
const (
	NamespaceMatchSelf          NamespaceMatchType = "Self"
	NamespaceMatchSelector      NamespaceMatchType = "Selector"
	NamespaceMatchSameLabels    NamespaceMatchType = "SameLabels"

> 	NamespaceMatchSameLabels    NamespaceMatchType = "NotSameLabels"

Reviewer 4: last const name is missing "Not". Though I might say "DifferentLabels" anyway?

)

```

#### AppliedTo
The `AppliedTo` field in Cluster-scoped network policies is what `Spec.PodSelector` field is to K8s NetworkPolicy spec,
as means to specify the target Pods that this cluster-scoped policy applies to.

> Since the policy is cluster-scoped, the `NamespaceSelector` field is required.

Reviewer 4: This contradicts what was written above.

Reviewer: ACK, will fix

An empty `NamespaceSelector` (namespaceSelector: {}) selects all Namespaces in the Cluster.

#### Namespaces
The `Namespaces` field complements `NamespaceSelector` in peers, as means to specify the Namespaces of
ingress/egress peers for cluster-scoped policies.
The scope of the Namespaces to be selected is specified by the matching strategy chosen.
For selecting Pods from specific Namespaces, the `Selector` scope works exactly as `NamespaceSelector`.
The `Self` scope is added to satisfy the specific needs for cluster-scoped policies:

__Self:__
This is a special strategy to indicate that the rule only applies to the Namespace for
which the ingress/egress rule is currently being evaluated upon. Since the Pods
selected by the ClusterNetworkPolicy `appliedTo` could be from multiple Namespaces,
the scope of ingress/egress rules whose `scope=self` will be the Pod's
own Namespace for each selected Pod.
Consider the following example:

> - Pods [a1, b1] exist in Namespace x, which has labels `app=a` and `app=b` respectively.
> - Pods [a2, b2] exist in Namespace y, which also has labels `app=a` and `app=b` respectively.

Reviewer 4: ```suggestion
- Pods [a1, b1], with labels `app=a` and `app=b` respectively, exist in Namespace x.
- Pods [a2, b2], also with labels `app=a` and `app=b` respectively, exist in Namespace y.
```

```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 100
  - appliedTo:
     namespaceSelector: {}
    ingress:
    - action: Allow
      from:
      - namespaces:

>           scope: self

Reviewer 4: ```suggestion
          scope: Self
```

the _same Namespace_ who has label app=b. Hence, the policy above allows
x/b1 -> x/a1 and y/b2 -> y/a2, but does not allow y/b2 -> x/a1 and x/b1 -> y/a2.

__SameLabels:__
This is a special strategy to indicate that the rule only applies to the Namespaces
which share the same label value. Since the Pods selected by the ClusterNetworkPolicy `appliedTo`
could be from multiple Namespaces, the scope of ingress/egress rules whose `scope=samelabels; labels: [tenant]`
will be all the Pods from the Namespaces who have the same label value for the "tenant" key.
Consider the following example:

- Pods [a1, b1] exist in Namespace t1-ns1, which has label `tenant=t1`.
- Pods [a2, b2] exist in Namespace t1-ns2, which has label `tenant=t1`.
- Pods [a3, b3] exist in Namespace t2-ns1, which has label `tenant=t2`.
- Pods [a4, b4] exist in Namespace t2-ns2, which has label `tenant=t2`.

```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 50
  - appliedTo:
     namespaceSelector:
       matchExpressions: {key: "tenant"; operator: Exists}
    ingress:
     - action: Pass
       from:
       - namespaces:

>           scope: samelabels

Reviewer 4: ```suggestion
          scope: SameLabels
```

the cluster who has a label key set as "tenant", traffic for all Pods in that Namespace
from all Pods in the Namespaces who has the same label value for key `tenant` is delegated to the Namespace
admins, i.e such traffic will not be subject to any CNP (`priority` > 50) rules and be evaluated by K8s NetworkPolicies.
Hence, the policy above delegates traffic from all Pods in Namespaces labeled `tenant=t1` i.e. t1-ns1 and t1-ns2,
to reach each other, to K8s NetworkPolicies, similarly traffic for all Pods in Namespaces labeled `tenant=t2`
i.e. t2-ns1 and t2-ns2, to talk to each other is delegated to K8s NetworkPolicies as well, however it does not
delegate traffic from any Pod in t1-ns1 or t1-ns2 to reach Pods in t2-ns1 or t2-ns2, such traffic is still subject
to CNP rules.

### Sample Specs for User Stories

#### Sample spec for Story 1: Deny traffic at a cluster level

```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 10
  - appliedTo:
     namespaceSelector: {}
    ingress:
     - action: Deny
       from:
       - namespaceSelector:
          matchLabels:
            isolatedByAdmin: true

> ```

Reviewer 4: So this says "for every pod in every namespace, disallow ingress from `isolatedByAdmin: true` namespaces". It would most likely be more efficient to flip it around and say "for every pod in every `isolatedByAdmin: true` namespace, denying all egress":

```
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 10
  - appliedTo:
      namespaceSelector:
        matchLabels:
          isolatedByAdmin: true
    egress:
     - action: Deny
```

> ```

Reviewer 4: That said, both the original policy and my proposed replacement are wrong; the policy applies to pods _inside_ the isolated namespace as well, preventing them from talking even to other pods in the same namespace, which does not appear to have been the intent. So you'd actually have to do:

```
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 10
  - appliedTo:
      namespaceSelector:
        matchLabels:
          isolatedByAdmin: true
    egress:
      - action: Pass
        to:
          namespaces:
            scope: Self
      - action: Deny
```
I think? (FTR, I got this wrong the first time by saying "Allow" instead of "Pass", which would have worked, but imposed more policy that I meant to.)

In the original form; it would be slightly more complicated:

```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 10
  - appliedTo:
      namespaceSelector:
        matchLabels:
          isolatedByAdmin: true
    ingress:
      - action: Pass
        from:
          namespaces:
            scope: Self
  - appliedTo:
      namespaceSelector: {}
    ingress:
      - action: Deny
        from:
          - namespaceSelector:
              matchLabels:
                isolatedByAdmin: true
```

> #### Sample spec for Story 2: Allow traffic at a cluster level

Reviewer 4: The CNP here doesn't include the DNS namespace from the diagram attached to story 2

Reviewer 2: Done

kind: ClusterNetworkPolicy
spec:
  priority: 30
  - appliedTo:
     namespaceSelector: {}
    ingress:
     - action: Allow
       from:
       - namespaceSelector:
           matchLabels:
             adminMonitoringNamespace: true
         workloadSelector:
           podSelector:
             matchLabels:
               app: adminMonitoringApp
```

#### Sample spec for Story 3: Explicitly Delegate traffic to existing K8s Network Policy

```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 20
  - appliedTo:
      namespaceSelector: {}
    ingress:
    - action: Pass
      from:
      - namespaceSelector:
          matchLabels:
            publicServiceTenant: true
        workloadSelector:
          podSelector:
            matchLabels:
              app: publicServiceFooBar
```

#### Sample spec for Story 4: Create and Isolate multiple tenants in a cluster

```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 50
  - appliedTo:
     namespaceSelector:
       matchExpressions: {key: "tenant"; operator: Exists}
    ingress:
     - action: Deny
       from:
       - namespaces:

>           scope: notsamelabels

Reviewer 4: ```suggestion
          scope: NotSameLabels
```

          labels:
            - tenant
```

Note: the above ClusterNetworkPolicy can also be written in the following fashion:
```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  priority: 50
  - appliedTo:
     namespaceSelector:
       matchExpressions: {key: "tenant"; operator: Exists}
    ingress:
     - action: Pass
       from:
       - namespaces:

>           scope: samelabels

Reviewer 4: ```suggestion
          scope: SameLabels
```

          labels:
            - tenant
     - action: Deny   # Deny everything else other than same tenant traffic
       from:
       - namespaceSelector: {}
```
The difference is that in the first case, traffic within tenant Namespaces will fall
through, and be evaluated against lower-priority ClusterNetworkPolicies, and then
NetworkPolicies. In the second case, the matching packet will skip all ClusterNetworkPolicy
evaluation (except for ClusterNetworkPolicy priority=baseline), and only match
against NetworkPolicy rules in the cluster. In other words, the second ClusterNetworkPolicy
specifies intra-tenant traffic must be delegated to the tenant Namespace owners.

#### Sample spec for Story 5: Cluster Wide Default Guardrails

```yaml
apiVersion: netpol.networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:

>   priority: baseline

Reviewer 4: ```suggestion
  priority: Baseline
```

    as per the `priority` set in CNP.
- e2e test cases must cover ingress and egress rules
- e2e test cases must cover port-ranges, named ports, integer ports etc
- e2e test cases must cover various combinations of `workloadSelector` in `appliedTo` and ingress/egress rules
- e2e test cases must cover various combinations of `namespaceSelector` in `appliedTo`
- e2e test cases must cover various combinations of `namespaces` in ingress/egress rules
  - Ensure that namespace matching strategies work as expected
- Add unit tests to test the validation logic which shall be introduced for cluster-scoped policy resources
  - Ensure that `self` field cannot be set along with `selector` within `namespaces`
  - Test cases for fields which are shared with NetworkPolicy, like `endPort` etc.
  - Ensure no two policies can be set with same `priority`
- Ensure that only administrators or assigned roles can create/update/delete cluster-scoped policy resources

### Graduation Criteria

#### Alpha to Beta Graduation

- Gather feedback from developers and surveys
- At least 2 CNI provider must provide the implementation for the complete set

>   of alpha features

Reviewer 4: We should discuss the numbers here. We used "2" for NetworkPolicy port ranges. I don't think we should be using the same criteria for something this large.

- At least 4 CNI providers must provide the implementation for the complete set
  of beta features
- More rigorous forms of testing
  — e.g., downgrade tests and scalability tests
- Allowing time for feedback
- Completion of all accepted "future work" items

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

#### Upgrade considerations

As such, the cluster-scoped policy resources are new and shall not exist prior
to upgrading to a new version. Thus, there is no direct impact on upgrades.

#### Downgrade considerations

Downgrading to a version which no longer supports cluster-scoped policy APIs
must ensure that appropriate security rules are created to mimick the cluster-scoped

> policy rules by other means, such that no unintended traffic is allowed.

Reviewer 4: And also that all _intended_ traffic _is_ allowed. eg, if there was a CNP that Allowed access to CoreDNS, they would need to make sure that everyone still had access without it

n/a

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

n/a

## Implementation History

- 2021-02-18 - Created initial PR for the KEP

## Drawbacks

Securing traffic for a cluster for administrator's use case can get complex.
This leads to introduction of a more complex set of APIs which could confuse
users.

## Alternatives

Following alternative approaches were considered:

### NetworkPolicy v2

A new version for NetworkPolicy, v2, was evaluated to address features and use cases
documented in this KEP. Since the NetworkPolicy resource already exists, it would be
a low barrier to entry and can be extended to incorporate admin use cases.
However, this idea was rejected because the NetworkPolicy resource was introduced
solely to satisfy a developers intent. Thus, adding new use cases for a cluster admin
would be contradictory. In addition to that, the administrator use cases are mainly
scoped to the cluster as opposed to the NetworkPolicy resource, which is `namespaced`.

### Empower, Deny, Allow action based CRD

Unlike the NetworkPolicy resource in which each rule represents an allowed

> traffic, ClusterNetworkPolicy will enable administrators to set `Empower`,

Reviewer 4: uh... no it won't...

This needs to be rewritten to reflect the fact that we _aren't_ implementing it this way

Reviewer: We wanted to mention these in the `Alternatives considered` section just to show it to the readers that, we have explored these other options but did not work out / was opposed by the community. I can try to make this section more concise, or simply redact it if you do think it's not needed at all.

Reviewer 4: Listing the old alternative ideas is good, but the text should be rewritten to talk about these approaches as things we _aren't_ doing (and explaining why we decided against them), rather than just preserving the old text verbatim and making it sound like we _are_ doing these things.

Reviewer 2: Done

`Empower` rules have the highest precedence, which shall only be used to provide
exceptions to deny rules. The `Empower` rules do not guarantee that the traffic
will not be dropped: it simply denotes that the packets matching those rules can bypass the
ClusterNetworkPolicy `Deny` rule evaluation.

ClusterNetworkPolicy `Deny` rules are useful for administrators to explicitly
block traffic from malicious clients, or workloads that poses security risks.
Those traffic restrictions can only be lifted once the `Deny` rules are deleted
or modified by the admin. In clusters where the admin requires total control over
security postures of all workloads, the `Deny` rules can also be used to deny all
incoming/outgoing traffic in the cluster, with few exceptions that's listed out
by `Empower` rules.

On the other hand, the `Allow` rules can be used to call out traffic in the cluster
that needs to be allowed for certain components to work as expected (egress to
CoreDNS for example). Those traffic should not be blocked when developers apply
NetworkPolicy to their Namespaces which isolates the workloads.

#### ClusterDefaultNetworkPolicy resource

> A ClusterDefaultNetworkPolicy resource will help the administrators set baseline

Reviewer 4: likewise

(you probably don't need all the detail either)

Reviewer 2: Done

}

type DefaultNetworkPolicyIngress/EgressRule struct {
	Ports            []networkingv1.NetworkPolicyPort
	OnlyFrom/OnlyTo  []DefaultNetworkPolicyPeer
}

type DefaultNetworkPolicyPeer struct {
	PodSelector  *metav1.LabelSelector
	// required if a PodSelector is specified
	Namespaces   *Namespaces
	IPBlock      *networkingv1.IPBlock
}
```

Most structs above are very similar to NetworkPolicy and quite self-explanatory.
One detail to notice is that in the DefaultNetworkPolicy Ingress/Egress rule spec,
the peers are created in a field named `OnlyFrom`/`OnlyTo`, as opposed to `To`/`From`
in ClusterNetworkPolicy. We chose this naming to better hint policy writers about
the isolation effect of DefaultNetworkPolicy on the workloads it applies to.

### Shared API Design
The following structs will be added to the `networking.k8s.io` API group and
shared between `ClusterNetworkPolicy` and `DefaultNetworkPolicy`:

```golang
type AppliedTo struct {
	// required if a PodSelector is specified
	NamespaceSelector   *metav1.LabelSelector
	// optional
	PodSelector         *metav1.LabelSelector
}

type Namespaces struct {
	// Self and Selector are mutually-exclusive

> 	Self       bool

Reviewer 7: Although both `self` and `selector` are about namespace, they are not very logically related. Would it be simpler to make them flat in `ClusterNetworkPolicyPeer`? I mean something like:
```
type ClusterNetworkPolicyPeer struct {
	PodSelector *metav1.LabelSelector

	// Select peer pods from namespaces selected by the namespaceSelector.
	NamespaceSelector *metav1.LabelSelector
        // Select peer pods from their own namespaces of the Pods which the ingress/egress rule is currently being evaluated upon.
	SelfNamespace bool

	IPBlock *networkingv1.IPBlock
}
``` 
Users who have been familar with `podSelector` and `namespaceSelector` can continue to define peer in the way they used to, and it's also easy to try out the new `selfNamespace` field in conjunction with `podSelector`.

Author: this is a good idea.. will propose this in the next meeting

user stories above.
1. Deny rules and, therefore, hierarchical enforcement of policy
2. Semantics for a cluster-scoped policy object that may include
   namespaces/workloads that have not been created yet.
3. Backwards compatibility with existing Kubernetes Network Policy API

### Non-Goals

Our mission is to solve the most common use cases that cluster admins have.
That is, we don't want to solve for every possible policy permutation a user
can think of. Instead, we want to design an API that addresses 90-95% use cases
while keeping the mental model easy to understand and use.

Additionally, this proposal is squarely focused on solving the needs of the
Cluster Administrator. It is not intended to solve:
1. Logging / error reporting for Network Policy
2. Kubernetes Node policies
3. New policy selectors (services, service accounts, etc.)
4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
   impact on workloads.

## Proposal

In order to achieve the two primary broad use cases for a cluster admin to
secure K8s clusters, we propose to introduce the following two new resources
under `networking.k8s.io` API group:
- ClusterNetworkPolicy

> - DefaultNetworkPolicy

Reviewer 8: @thockin can we do this? networking.k8s.io is in v1, is it possible to add new resources directly to a v1 group?

Reviewer 4: versions are under groups. `networking.k8s.io` has `v1` and `v1beta1` versions right now. The new types would (presumably) go under `kubernetes/api/networking/v1alpha1/`

Reviewer 8: the way I read this is that those types are removed and you can't bring them back
https://kubernetes.io/docs/reference/using-api/deprecation-guide/#networkpolicy-v116
actually, unrelated but it seems we no longer server any beta in networking.k8s.io since 1.22 https://kubernetes.io/docs/reference/using-api/deprecation-guide/#ingress-v122

Reviewer 3: But this considering the NetworkPolicy object. As their intent here is a new object type, IMO there's no big deal. It's going just to be a networking.k8s.io/v1alpha1 -> ClusterNetworkPolicy.

I guess it would be something like adding to https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/networking/types.go the specs, and the rest of the code https://github.com/kubernetes/kubernetes/tree/master/pkg/apis/networking/v1alpha1 (which does not exists)

As a cluster admin, I want to explicitly deny traffic from certain source IPs
that I know to be bad.

Many admins maintain lists of IPs that are known to be bad actors, especially
to curb DoS attacks. A cluster admin could use ClusterNetworkPolicy to codify
all the source IPs that should be denied in order to prevent that traffic from
accidentally reaching workloads. Note that the inverse of this (allow traffic
from well known source IPs) is also a valid use case.

#### Story 2: Funnel traffic through ingress/egress gateways

As a cluster admin, I want to ensure that all traffic coming into (going out of)
my cluster always goes through my ingress (egress) gateway.

It is common practice in enterprises to setup checkpoints in their clusters at
ingress/egress. These checkpoints usually perform advanced checks such as
firewalling, authentication, packet/connection logging, etc.
This is a big request for compliance reasons, and ClusterNetworkPolicy can ensure
that all the traffic is forced to go through ingress/egress gateways.

#### Story 3: Isolate multiple tenants in a cluster

As a cluster admin, I want to isolate all the tenants (modeled as Namespaces)

> on my cluster from each other by default.

Reviewer 9: change to (modeled as individual Namespaces or closed groups of co-operating namespaces)

Many enterprises are creating shared Kubernetes clusters that are managed by a
centralized platform team. Each internal team that wants to run their workloads
gets assigned a Namespace on the shared clusters. Naturally, the platform team
will want to make sure that, by default, all intra-namespace traffic is allowed
and all inter-namespace traffic is denied.

#### Story 4: Enforce network/security best practices

As a cluster admin, I want all workloads to start with a baseline network/security
model that meets the needs of my company.

A platform admin may want to factor out policies that each namespace would have
to write individually in order to make deployment and auditability easier.
Common examples include allowing all workloads to be able to talk to the cluster
DNS service and, similarly, allowing all workloads to talk to the logging/monitoring

> pods running on the cluster.

Reviewer 4: But it _doesn't_ justify the existence of DNP; that's my point. It starts out sounding like you're going to be talking about DNPs, but then the example you give is implemented with CNPs, and couldn't be implemented with DNPs.

> pods running on the cluster.

Reviewer 5: Yeah, I agree with Dan - I'd expect this to be a CNP.

Maybe I am missing it, but I'm not seeing a use-case that justifies the existence of DNPs?

> pods running on the cluster.

Author: correct. i think we are all on the same page here, in that this example is not a correct justification of DNP. I will sync with @gjsj and update the example

> pods running on the cluster.

Reviewer: > These user-stories are not really resonating for me. Or they are not complete anyway. Like the cross-NS requirements really need to explore both "...and the user can override it" as well as "...and the user can not override it". I think both are useful

Does that mean you feel like the "...and the user can override it" part should be solved by a DNP or no?

> pods running on the cluster.

Reviewer 6: I mean there are use-cases where the rule is soft (e.g. by default they can't cross-namespaces, but they are allowed to enable it) and there are cases where the rule is hard (coke can never enable access to pepsi, and vice-versa)

#### Namespaces
The `Namespaces` field replaces `NamespaceSelector` in NetworkPolicyPeer, as
means to specify the Namespaces of ingress/egress peers for cluster-scoped policies.
For selecting Pods from specific Namespaces, the `Selector` field under `Namespaces`
works exactly as `NamespaceSelector`. The `Self` field is added to satisfy the
specific needs for cluster-scoped policies:

__Self:__ An optional field, which evaluates to false by default.
When `self: true` is set, no selectors can be present concurrently. This is a
special keyword to indicate that the rule only applies to the Namespace for
which the ingress/egress rule is currently being evaluated upon. Since the Pods
selected by the ClusterNetworkPolicy appliedTo could be from multiple Namespaces,
the scope of ingress/egress rules whose `namespace.self=true` will be the Pod's
own Namespace for each selected Pod.
Consider the following example:

- Pods [a1, b1] exist in Namespace x, which has labels `app=a` and `app=b` respectively.
- Pods [a2, b2] exist in Namespace y, which also has labels `app=a` and `app=b` respectively.

```yaml

> apiVersion: networking.k8s.io/v1alpha1

Reviewer 8: I don't have enough knowledge about this, @liggitt is it possible to enable an api group version that was already deprecated, in this case `networking.k8s.io/v1alpha1` ?

kind: DefaultNetworkPolicy
spec:
  appliedTo:
    namespaceSelector: {}
  ingress:
    - onlyFrom:
      - namespaces:
          self: true
        podSelector:
          matchLabels:
            app: b
```

The above DefaultNetworkPolicy should be interpreted as: for each Namespace in
the cluster, all Pods in that Namespace should only allow traffic from Pods in
the _same Namespace_ who has label app=b. Hence, the policy above allows
x/b1 -> x/a1 and y/b2 -> y/a2, but denies y/b2 -> x/a1 and x/b1 -> y/a2.

#### IPBlock

The `ClusterNetworkPolicyPeer` and `DefaultNetworkPolicyPeer` both allow the
ability to set an `IPBlock` as a peer. The usage of this field is similar to
how it is used in the NetworkPolicyPeer. However, we should also explicitly
note that the IPBlock set in this field could belong to the cluster locally, or

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 4: _The first user story in this KEP_ is "As a cluster admin, I want to explicitly deny traffic from certain source IPs that I know to be bad." If you leave `ipBlock` behavior implementation-defined, you are not solving that user story. This is a crucial feature. We can't just punt on it.

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 2: +1 here, one of the major problems I've seen with Network Policy is leaving certain behaviors of selectors (like IPBLOCK) as implementation Specific. This results in many caveats across CNI's.... If the behavior is something that can easily be clearly defined I think it should be.

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 5: > the source IP seen by the CNI implementation is what will be considered

I think it needs to be stronger than this. CNI isn't necessarily Service aware, but NetworkPolicy as a Kubernetes API might be expected to be. I think the NodePort example really gets to the crux of this issue. We have previously talked about this as whatever source IP the pod itself sees - that's useful for implementations, and not _entirely_ broken for users, but a more usable API would maybe extend a little bit further - i.e., the source IP as it enters the Kubernetes cluster. It's probably not fair to say NetworkPolicy should be aware of devices outside of the Kubernetes cluster itself, though.

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 10: +1 to what Casey said here. While it is not possible to enforce the policy on original source IP if the SNAT happens outside the cluster (ex: cloud-provider Load Balancer), we can certainly define the API to consider the source IP as it enters kubernetes cluster.

Considering a specific traffic flow for node port,`EXTERNAL-IP -> NODE-IP1:NODE-PORT` that gets rewritten to NODE-IP1 -> POD2(on NODE2):PORT. In this case, the source IP is considered to be NODE-IP1 whereas the actual IP address is an external IP address. This not only allows all external IP addresses that are allowed by cloud-provider firewall but also interfere with node based policies that we introduce in future.

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 5: For completeness, the reason we haven't been able to take such a strong stance to-date is because network policy works in terms of _pods_ and not services. 

Take for example a service that selects both pod A and pod B, but each of those pods has different network policy applied. Most implementations will need to understand which of those pods will be the ultimate destination in order to apply the correct policy. At the point that traffic enters the cluster, that may not be knowable (e.g., since kube-proxy will arbitrarily assign the traffic to one of those pods).

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 10: Yeah I agree that it will introduce additional complexities for CNI implementers.

I think the enforcement can happen at two places,
1. The node that first received the packet
2. The node that hosts the backend pod

The service resolution happens on the node that first received the packet. If the policy enforcement is to happen after the service resolution on this node, then we know about the backend pod.

If the enforcement happen on the node that hosts the backend pod, the CNI implementer need to use global conntrack state or some other mechanism to identify the original source IP.

In both cases, there seems to be a way to know about the backend pod depending on where in the stack the policy is enforced.

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 5: @abhiraut so I want to make it clear that the above was more of a "this is probably what is best" type discussion.

I believe we have already defined the `ipBlock` semantics for the existing NetworkPolicy API to be "as seen by the selected pod" - although I know this was discussed when originally implementing the API, this doesn't seem to be written down anywhere, and so I can't prove it. .  Are you suggesting with this KEP that we change the semantic for both? Or that we have different behaviors for CNP vs regular NP?

I think the former would be great, if we can do it without breaking API compatibility contracts. I think the latter would be a bad outcome.

More generally though, I think re-deifning what `ipBlock` means is arguably out of scope for this KEP.

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 4: > I believe we have already defined the `ipBlock` semantics for the existing NetworkPolicy API to be "as seen by the selected pod" - although I know this was discussed when originally implementing the API, this doesn't seem to be written down anywhere, and so I can't prove it.

It's in https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-selectors.

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 4: > I'm not at all convinced that "as it enters the Kubernetes cluster" is realistic.

Possibly, but if so, then this KEP does not solve the problems of the users in its User Stories, and so should be rejected (or at least, rewritten to not claim to be solving those problems, in which case parts of it may no longer make sense as part of a solution that is only trying to solve the remaining user stories, etc).

> something about trusting the various ingress mechanisms to all be covered makes me uneasy.

When I was trying to write up a KEP for egress-only admin-focused network policy a while back ([ClusterEgressFirewall](https://github.com/danwinship/enhancements/blob/cluster-egress-firewall/keps/sig-network/20190917-cluster-egress-firewall.md)) I said:

> Unlike with NetworkPolicy, which is expected to be implemented entirely by the network plugin, ClusterEgressFirewall will be a shared responsibility; any component of Kubernetes that provides a way for packets to exit the cluster (eg, secondary network interfaces) must ensure that egress firewall rules will be obeyed by exiting traffic. Any component that has the effect of obscuring the origin of packets (eg, proxies or load balancers) must ensure that they cannot be used to bypass egress firewall rules.

(The "secondary network interfaces" part is probably definitely out of scope for ClusterNetworkPolicy but the various issues around using cloud load balancers to bypass policy rules still apply.)

> could be cluster external. For example, the peer could be set with a subnet

Reviewer 6: if I have a proxyful LB implementation, and I set up an ipBlock rule, the proxy IP must be included in the rule, but proxy IPs are not a useful unit to block.  Saying that the proxyful LB must respect NetworkPolicy is maybe appropriate but is a huge net-new requirement and we'd have to think really hard about whether it was feasible or not.

Or we simply don't - if you are exposing something "externally" then there's a different expression of policy needed (e.g. pod selector does not make sense)

As a cluster admin, I want to explicitly deny traffic from certain source IPs
that I know to be bad.

```yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
metadata:
  name: deny-bad-ip
spec:
  appliedTo:
    # if there's an ingress gateway in the cluster, applying the policy to
    # gateway namespace will be sufficient
    namespaceSelector: {}
  ingress:
    - action: Deny
      from:
      - ipBlock:
          cidr: 62.210.0.0/16  # blacklisted addresses
```

#### Story 2: Funnel traffic through ingress/egress gateways
As a cluster admin, I want to ensure that all traffic coming into (going out of)
my cluster always goes through my ingress (egress) gateway.

```yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
metadata:
  name: ingress-egress-gateway
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        type: tenant  # assuming all tenant namespaces will be created with this label
  ingress:
    - action: Deny
      from:
      - ipBlock:
          cidr: 0.0.0.0/0
      except:
      - namespaces:
          self: true
      - namespaces:
          selector:
            matchLabels:

>               kubernetes.io/metadata.name: dmz  # ingress gateway

Reviewer: Thanks for the suggestion! Will update this

  egress:
    - action: Deny
      to:
      - ipBlock:
          cidr: 0.0.0.0/0
      except:
      - namespaces:
          self: true
      - namespaces:
          selector:
            matchLabels:
              kubernetes.io/metadata.name: istio-egress  # egress gateway
```

__Note:__ The above policy is very restrictive, i.e. it rejects ingress/egress
traffic between tenant Namespaces and `kube-system`. For `coredns` etc. to work,
`kube-system` Namespace or at least the `coredns` pods needs to be added into

> the Deny `except` list.

Reviewer: I agree this is indeed a better spec for the usecase. But going off the discussion here https://github.com/kubernetes/enhancements/pull/2522#discussion_r601718058, how would we write such policy if there were no `except` field? Would the cluster admin need to manually put the cluster CIDR in the `ipBlock.except`?

> the Deny `except` list.

Reviewer 4: Agreed, this use case couldn't easily be described using (the current version of this KEP minus `except`) or even (the current version of this KEP minus `except` plus explicit priority levels).

But there are other ways it might be implemented. eg, if `ipBlock` was understood to only refer to cluster-external IPs, not cluster-internal IPs, then you wouldn't need an `except`. (@tgraf and I were just discussing that...)

- e2e test cases must cover various combinations of `podSelector` in `appliedTo` and ingress/egress rules
- e2e test cases must cover various combinations of `namespaceSelector` in `appliedTo`
- e2e test cases must cover various combinations of `namespaces` in ingress/egress rules
  - Ensure that `except` field works as expected
  - Ensure that `self` field works as expected
- Add unit tests to test the validation logic which shall be introduced for cluster-scoped policy resources
  - Ensure that `self` field cannot be set along with `selector` within `namespaces`
  - Test cases for fields which are shared with NetworkPolicy, like `ipBlock`, `endPort` etc.
- Ensure that only administrators or assigned roles can create/update/delete cluster-scoped policy resources

### Graduation Criteria

#### Alpha to Beta Graduation

- Gather feedback from developers and surveys
- At least 1 CNI provider must provide the implementation for the complete set
  of alpha features
- Evaluate "future work" items based on feedback from community

#### Beta to GA Graduation

- At least 2 CNI providers must provide the implementation for the complete set

>   of alpha features

Reviewer 4: presumably should say "beta" there not alpha.

I'm not sure that "1" is a good value for the alpha-to-beta transition. I _am_ sure that "2" is not a good value for the beta-to-GA transition. This is a major feature. We want more buy-in than that before calling it GA.

- More rigorous forms of testing
  — e.g., downgrade tests and scalability tests
- Allowing time for feedback
- Completion of all accepted "future work" items

#### Removing a Deprecated Flag

- Announce deprecation and support policy of the existing flag
- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
- Address feedback on usage/changed behavior, provided on GitHub issues

> - Deprecate the flag

Reviewer 4: There are no flags involved here. This section does not apply; you can just delete it.

new API calls will be made.

###### Will enabling / using this feature result in introducing new API types?

Enabling this feature will introduce new API types as described in the [design](#design-details)
section. The supported number of objects per cluster will depend on the individual
CNI providers who will be responsible to provide the implementation to realize
these resources.

###### Will enabling / using this feature result in any new calls to the cloud provider?

n/a

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

n/a

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

n/a

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

n/a

## Implementation History

- 2021-02-18 - Created initial PR for the KEP

## Drawbacks

Securing traffic for a cluster for administrator's use case can get complex.
This leads to introduction of a more complex set of APIs which could confuse
users.

## Alternatives

> Following alternative approaches were considered:

Reviewer 8: Have you evaluated the option to implement it as a CRD instead of a core type, ala Gateway API? https://github.com/kubernetes-sigs/gateway-api#kubernetes-gateway-api

Reviewer 8: Is not the norm, but in-tree API objects has much more requirements and the development process is much slower. Having people that went through this thought process is at least worth asking

Reviewer 5: I would have suggested that we start with a CRD for sure.

Author: sounds good. I am thinking a more generic repo `network-policy-api` under `kubernetes-sigs` to evolve related CRDs, for example, CNPs, the v2 NP CRD etc. Let us kick start a discussion around those lines and we will make changes here accordingly and in the mean time, we can continue reviewing the spec

Reviewer 2: +1 @abhiraut I think that will be the most efficient path forward

Reviewer 6: I think this should land as a built-in because it makes sense to be in the same group as NetworkPolicy.  We can start with a CRD in a different group.

# KEP-2091: Add support for ClusterNetworkPolicy resources

> <!-- toc -->

Author: done

user stories above.
1. Deny rules and, therefore, hierarchical enforcement of policy
2. Semantics for a cluster-scoped policy object that may include
   namespaces/workloads that have not been created yet.
3. Backwards compatibility with existing Kubernetes Network Policy API

### Non-Goals

Our mission is to solve the most common use cases that cluster admins have.
That is, we don't want to solve for every possible policy permutation a user
can think of. Instead, we want to design an API that addresses 90-95% use cases
while keeping the mental model easy to understand and use.

Additionally, this proposal is squarely focused on solving the needs of the
Cluster Administrator. It is not intended to solve:
1. Logging / error reporting for Network Policy
2. Kubernetes Node policies
3. New policy selectors (services, service accounts, etc.)
4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
   impact on workloads.

## Proposal

In order to achieve the two primary broad use cases for a cluster admin to
secure K8s clusters, we propose to introduce the following two new resources
under `networking.k8s.io` API group:
- ClusterNetworkPolicy

> - DefaultNetworkPolicy

Reviewer: I think we can all agree that both resources speak to the same persona, and that's why we put them in the same KEP. We also started out with a single ClusterNetworkPolicy resource, and defined a `defaultAllow` action. But there was one problem: the rule semantics are different between `defaultAllow` and `allow`/`deny`.  `defaultAllow` won't make sense if there are no implicit isolation enforced, since all Pods should be able to communicate with each other already, without policies. `allow`/`deny` however, shouldn't create any implicit rules and can be read as-is. 

This caused a lot of confusion when we presented it in sig-net bi-weekly (https://www.youtube.com/watch?v=gzLJNNBTQ8w and https://www.youtube.com/watch?v=t7rRF4W-dXg), and led us to think maybe it is better to define separate CRDs for those two intents. 

The `alternatives` section below does explore several possibilities in terms of single CRD (in addition to what you proposed). We could continue the discussion there.

> - DefaultNetworkPolicy

Reviewer 5: Naively, what about a sincle ClusterNetworkPolicy API with something like this:

```
spec:
  canBeOverridden: true | false
```

That's the main difference between the CNP and DNP, right? I don't know if that's actually better... but if the goal is to have fewer `Kind`s then that might do it.

> - DefaultNetworkPolicy

Author: the boolean field is one of the alternatives proposed towards the end of this KEP, along with other alternatives which are all single CRD approach.

### ClusterNetworkPolicy resource

A ClusterNetworkPolicy resource will help the administrators set strict
security rules for the cluster, i.e. a developer CANNOT override these rules
by creating NetworkPolicies that applies to the same workloads as the
ClusterNetworkPolicy does.

Unlike the NetworkPolicy resource in which each rule represents a allowed
traffic, ClusterNetworkPolicy will enable administrators to set `Allow` or
`Deny` as the action of each rule. ClusterNetworkPolicy rules should be read

> as-is, i.e. there will not be any implicit isolation effects for the Pods

Reviewer 5: On its own this is fine, but it might be confusing to have two/three parallel APIs that have fairly important, subtle differences like this.

Not necessarily an issue but something to think about. In Calico's equivalent we have used an explicit "Pass" action so policy authors can write an explicit rule to send traffic to the developer policies, while maintaining the same general default behaviors.

Reviewer 5: We might want to consider a change of ClusterNetworkPolicy to something like ClusterAdminNetworkPolicy, etc. The behavior is different enough from a NP that it's not simply the cluster-scoped version of it, it's an entirely different thing.

rules with action `Deny` in the cluster combined) should be evaluated before
aggregated ClusterNetworkPolicy `Allow` rules, followed by aggregated
NetworkPolicy rules in all Namespaces. As such, the `Deny` rules have the
highest precedence, which prevents them to be unexpectedly overwritten.

ClusterNetworkPolicy `Deny` rules are useful for administrators to explicitly
block traffic from malicious clients, or workloads that poses security risks.
Those traffic restrictions can only be lifted once the`Deny` rules are deleted
or modified. On the other hand, the `Allow` rules can be used to call out
traffic in the cluster that needs to be allowed for certain components to work
as expected (egress to CoreDNS for example). Those traffic should not be blocked
when developers apply NetworkPolicy to their Namespaces which isolates the workloads.

### DefaultNetworkPolicy resource

A DefaultNetworkPolicy resource will help the administrators set baseline security
rules for the cluster, i.e. a developer CAN override these rules by creating
NetworkPolicies that applies to the same workloads as the DefaultNetworkPolicy does.

> DefaultNetworkPolicy works just like NetworkPolicy except that it is cluster-scoped.

Reviewer 5: Yeah, I agree, although I would suggest ClusterDefaultNetworkPolicy (in case we ever find a need for a namespaced DefaultNetworkPolicy). It would match the standard naming convention of Foo and ClusterFoo used elsewhere.

When workloads are selected by a DefaultNetworkPolicy, they are isolated except
for the ingress/egress rules specified. DefaultNetworkPolicy rules will not have
actions associated -- each rule will be an 'allow' rule.

> Aggregated NetworkPolicy rules will be evaluated before aggregated DefaultNetworkPolicy rules.

Reviewer 5: Not really a suggestion but rather a comment - this is somewhat aligned with how we thought about NetworkPolicy when we first designed it. Namely that NetworkPolicies are an override to the "cluster default" behavior, except that previously the cluster default behavior was implicitly defined (normally allow-all, but not strictly always). This provides a way to set it explicitly.

If a Pod is selected by both, a DefaultNetworkPolicy and a NetworkPolicy, then

> the DefaultNetworkPolicy's effect on that Pod becomes obsolete.

Reviewer 5: It feels a bit counter-intuitive to me that the NetworkPolicy "default deny" gets implemented before the DefaultNetworkPolicy - doesn't this render DNPs somewhat useless? I think I might not be grokking what the purpose of DNP is.

Reviewer 5: I would have expected:

1. NetworkPolicy rules
2. DefaultNetworkPolicy rules
3. Default deny

Reviewer: We gave this a bit of thought before opening the KEP, and I believe the reason we wanted NetworkPolicy default deny to have higher priority over DefaultNetworkPolicy rules is that, Namespace users will not be able to overwrite DefaultNetworkPolicy rules otherwise (or maybe this is what's expected in most cases based on previous comments?), so that statement about DefaultNetworkPolicy rules become obsolete when user apply NetworkPolicy becomes a lie

Reviewer 5: The implicit default-deny-on-selected behavior being different across the different policy types feels like a major area for confusion to me. It seems to me that the most intuitive behavior would be that whenever _any_ type of policy selects a pod, it's default behavior goes from allow to deny.

Then, it's simply a matter of rule ordering.

1. cluster-exceptions
2. cluster-deny
3. cluster-allow
4. namespaced NP rules.
5. default-deny.

This does preclude writing more "complex" DefaultNetworkPolicies, but I have yet to see a real use-case for a DNP. The closest thing is "default to deny anything except within my namespace", but I think that can be implemented without DNP.

e.g., 

1. A CNP that selects everything, allows nothing (thus creating a default-deny for all pods).
2. a NP in each namespace that allows within the NS. This could, of course, be edited or removed by the users within that NS, but that's no different than the "overridding" mechanics present in the current DNP proposal.

Reviewer: >  It seems to me that the most intuitive behavior would be that whenever _any_ type of policy selects a pod, it's default behavior goes from allow to deny.

We decided that the CNP actions should be 'as-is' because the semantics is less implicit this way - it has a couple of major benefits:
- Each allowed/dropped traffic matching a CNP will be attributed to a specific rule, rather than implied by some isolation effects
- It makes the meaning of the CNP actions more 'pure'. If these actions come with default isolation, then a standalone CNP `Deny` rule won't make any sense: isolate ingress for workload xxx and deny ingress from yyy specifically?

> A NP in each namespace that allows within the NS. This could, of course, be edited or removed by the users within that NS

In this KEP use think the Namespace (as tenant boundary) isolation should be the persona of a cluster admin rather than a namespace user. Of course the same traffic enforcements can be achieved by NetworkPolicies, but as you pointed out, the developers in each Namespace would have to maintain essentially the same policy, which should come as a cluster default for them.

Reviewer 5: I tend to consider consistency between similar APIs to be one-of, if not _the_ top consideration when building a new API. If this were _replacing_ NetworkPolicy, we'd have a bit more freedom IMO. But this is an API that is meant to work alongside NetworkPolicy, and so having subtle behavioral/syntactic/etc. differences is an easy way to make the APIs harder for users to get their heads around.

Reviewer: I do agree that consistency is a very important factor. As a matter of fact, for the CRDs that we plan to introduce to the new repo (which would host the ClusterNetworkPolicy resources and other improvements to NetworkPolicy), consistency is our main concern. This is why @abhiraut  @astoycos @rikatz and I believe CNP might be a good start for us to get rid of the implicit isolation behavior, even though it will need to co-exist with the current NetworkPolicy for a while. If new semantics makes sense, then it can become the guiding principle for the new CRDs, which eventually _can_ replace the current NetworkPolicy

Reviewer 2: @caseydavenport Network Policy Was originally designed to express the Developer's Intent, while the main persona in this proposal is the administrator so point being.....shouldn't 90% of design choices be done with the main user in mind? If so then why are we following precedents set by Network Policy, it's like trying to put a square peg in a round hole. 

We know Network Policy is here to stay, and is a great tool for the developer to apply security policies, however I would argue it is not a great tool for an administrator for a number of reasons, one of them being the implicit isolation behavior. 

I guess my main point is that we're starting the new `network-policy-api` repo to house the CNP and a possible new object around Network Policy (I hate to call it V2 :) in a manner that won't break any existing API.  Therefore, If the community decides that the new personas/ user stories for such objects are not in line with Existing NP then is there really anything holding us back from breaking away from the conventions used previously?

Reviewer 5: > shouldn't 90% of design choices be done with the main user in mind? If so then why are we following precedents set by Network Policy, it's like trying to put a square peg in a round hole.

While we talk about two distinct personas in the proposal, in the real-world these roles may be implemented by the same person or people within a business. Or they might even be split among more than two different people/groups.

> I hate to call it V2 ... Therefore, If the community decides that the new personas/ user stories for such objects are not in line with Existing NP then is there really anything holding us back from breaking away from the conventions used previously?

Yep, this is exactly what I am trying to explain. We're not building a v2. We're not replacing the existing API. We're augmenting it, and the new API(s) should behave consistently and intuitively with the APIs that already exist in v1. 

Even if one makes the argument that the existing default-deny mechanics are confusing, adding another API that behaves differently won't fix that, and will instead just make the set of APIs even more confusing.

If we ever do build a v2 NetworkPolicy, then that should be its own effort separate from this ClusterNetworkPolicy proposal, and that would be the time we should consider making changes across the board to NP / CNP / DNP / whatever NP APIs exist. But we would still aim for the set of v2 APIs to be consistent with each other. 

I'm not the only person who was caught off-guard by this unexpected difference: https://github.com/kubernetes/enhancements/pull/2522#discussion_r617282128

The responses I received above talk mostly in vague terms like "API purity", "less implicit", etc. Is there a cut-and-dry use-case that necessitates this? Because that's a lot more convincing of an argument. In other words, we need to justify making this behavior change with a use-case rather than the other way around.

Reviewer 4: > The implicit default-deny-on-selected behavior being different across the different policy types feels like a major area for confusion to me

I agree with this 100%, though I disagree on the fix; I think we should deprecate NetworkPolicy and replace it with a new ApplicationNetworkPolicy that is mostly similar, but has only explicit deny semantics.

Implicit default-deny-on-selected is the single most terrible decision sig-network has ever made (which is understandable given that we made it under extreme time pressure after discovering a day or two before feature freeze that we weren't going to be allowed to turn our Namespace annotation into a Namespace field).

> It seems to me that the most intuitive behavior would be that whenever _any_ type of policy selects a pod, it's default behavior goes from allow to deny.
> 
> Then, it's simply a matter of rule ordering.
> 
>     1. cluster-exceptions
>     2. cluster-deny
>     3. cluster-allow
>     4. namespaced NP rules.
>     5. default-deny.

That means there's no way to create a policy saying "allow ingress from prometheus to all namespaces" without causing the entire cluster to become isolated. There is no way to just say "I want to allow prometheus to reach all pods". You have to say "I want to allow prometheus to reach all pods, and I also want to block all other traffic".

Reviewer 5: > I agree with this 100%, though I disagree on the fix; I think we should deprecate NetworkPolicy and replace it with a new ApplicationNetworkPolicy that is mostly similar, but has only explicit deny semantics.

Yep, I think this would also be a fine approach, ans is basically aligned with what I was saying above - either we need to build something that plays nice with NetworkPolicy, or we need to build something entirely new that replaces NetworkPolicy, but in either case we need consistency. 

> That means there's no way to create a policy saying "allow ingress from prometheus to all namespaces" without causing the entire cluster to become isolated.

Yeah, I had this realization the other day as well. I think this is probably why the proposal is the way that it is currently. I suspect there are other ways to resolve this, but they might be even less desirable.

Reviewer 4: > something entirely new that replaces NetworkPolicy

FWIW, we could define "ApplicationNetworkPolicy" in a way that would still be compatible with NetworkPolicy. Just give it explicit Allow and Deny rules, where the Allows have priority over the Denies, and then say that a NetworkPolicy that says

    ingress:
      - from:
          - X
          - Y
          - Z

is equivalent to an ApplicationNetworkPolicy that says

    ingress:
      - action: Allow
          - from:
              - X
              - Y
              - Z
      - action: Deny
          - from:
              - ipBlock:
                  cidr: 0.0.0.0/0

Then it's not difficult to support both new ApplicationNetworkPolicies and legacy NetworkPolicies at the same time, and it's always clear how legacy NetworkPolicies interact with CNPs/DNPs, etc (although if we were going to do this then this would presumably involve changes to how CNP and DNP were defined too.)

Reviewer 2: >While we talk about two distinct personas in the proposal, in the real-world these roles may be implemented by the same person or people within a business. Or they might even be split among more than two different people/groups.

I agree, but at the same time we need to design for one of them... Trying to "optimize" a design for two different personas is what I think has been causing some of the confusion in this KEP. 

Overall the question of what API this fits into definitely needs to be solidified.  

@danwinship I like that idea, there was actually some discussion about having the same sort of thing in CNP (i.e two actions for a CNP/DNP  with allows overriding denys) and I think it's something that _could_ be fit into the existing API. 

@Dyanngg This ^ is sort of what I was referring to in the meeting today. The decision about how actions will work with CNP (i.e the 3 actions, only allow/deny, rules with priority, etc. ) Will ultimately be a major factor in the question of are we trying to "augment" the existing API or begin forming a new set of "v2" APIs. Specifically something like what Dan is proposing could "fit" into the existing API while something like actions with configurable priorities may not. 

Hopefully after further analysis of the main personas and core user stories it will be a bit easier to draw some conclusions here.

In this case, the traffic allowed will be solely determined by the NetworkPolicy.

> ### Precedence model

Reviewer 4: DNP is also supposed to allow for overriding the baseline accept/deny policy. That is, kubernetes clusters by default have the rule "pods that are selected by no NetworkPolicies allow everything", but with DNP you can change that to "pods that are selected by no NetworkPolicies deny everything" or even "pods that are selected by no NetworkPolicies allow from other pods in the same namespace but not from pods in other namespaces". You can't implement that with your suggestion.

(The counterargument would be that this is really a totally different kind of thing from "I want to block pods from talking to A.B.C.D" / "I want to force pods to accept connections from `monitoring-namespace`", and should be defined as a separate object, rather than trying to squish both pieces of functionality together.)

> ### Precedence model

Reviewer: > DNP is also supposed to allow for overriding the baseline accept/deny policy. 
Thanks for calling it out @danwinship. To your counter-argument, we also thought about making this baseline intent a config option for the cluster for example, but it will be hard for admins to fine-tune it and dynamically control the baseline postures.

> ### Precedence model

Author: we incorporated @tgraf suggestion of `exceptions` (named `authorize`).

for the `default` action, there was quite a bit of push back when it was first proposed in sig-network, hence keeping it as is in the form of 2 CRDs for now

> ### Precedence model

Reviewer 5: Yeah that was my feeling about DNP as well and I believe I mentioned it somewhere else. I think the reasoning is to avoid needing to rubber-stamp resources into each namespace. But like you said I'm not sure if that's justification for a new resource. It can just be scripted with a controller or something.

> ### Precedence model

Reviewer 4: > As it is defined (user NetPol totally deactivates it), how is it different than just putting a NetworkPolicy resource into the user's namespace?

Because you can write a DNP that sets the baseline policy for all namespaces, including namespaces that haven't yet been created, but there's no way to ensure that a NetworkPolicy gets automatically created in every newly-created namespace. And there's a big difference between "all Namespaces are isolated by default" and "please remember to make all of your Namespaces isolated after you create them!"

Additionally, this preserves the distinction between administrator intent and developer/deployer intent; NetworkPolicies are for developers to use, CNPs and DNPs are for administrators to use.

> ### Precedence model

Reviewer 9: The namespace/ tenant isolation requirement can be either admin-controlled or user-controlled in practice:  Either the (a) cluster admin knows and enforces the grouping of "co-operating namespaces" which make up a tenant or (b) the cluster admin lets the users/ tenants fully control whichever set of namespaces they want to group/ allow communication and has no direct visibility or control.   For (a), the namespace group isolation should be enforced via CNP and not via DNP. For (b) If the cluster admin can only inject initial config but has no easy way of verifying what parts of the DNP are being over-riden by the namespace NPs, then the cluster admin probably wont care to use the DNP as a tool. For a true self-service namespace grouping/ multitenancy model, there need to be other ways to let the cluster admin have visibility and predictability while retaining self-service ability for the end users.

It seems the combination of an admin-generated DNP + user generated NP has unpredictable behavior for the cluster admin so they wont care to use it very much. So either not have the DNP at all or decide that once a user provisions an NP, then all DNPs in that namespace become dont cares).

Also I would think that app developers would want to write cluster independent network policies where possible to ensure portability hence cant assume that DNP in cluster A would be the same as DNP in cluster B so would anyway cover all their requirements via NPs rather than relying on the cluster-dependent combination of NP + DNP.

<ended up being a longer comment than I'd have liked :-)>

> ### Precedence model

Reviewer 9: Also we may want to revisit and consider having simple priority ordering (sequence numbers etc) instead of various combinations of Authorize, Except, AlwaysAllow, Default etc with implicit priorities based on action type.  Also note that we would want to be able to add future actions like "mirror packet" or "log packet" and priority of the rule should not be overloaded and hard coded into the action type.

> ### Precedence model

Reviewer 4: > Either the (a) cluster admin knows and enforces the grouping of "co-operating namespaces" which make up a tenant or (b) the cluster admin lets the users/ tenants fully control whichever set of namespaces they want to group/ allow communication and has no direct visibility or control. For (a), the namespace group isolation should be enforced via CNP and not via DNP. For (b) If the cluster admin can only inject initial config but has no easy way of verifying what parts of the DNP are being over-riden by the namespace NPs, then the cluster admin probably wont care to use the DNP as a tool.

There are admins who want "I want to lock down everything my way and not let users bypass it" and they can use CNP. But there are also lots of admins who want "I want the default to be locked-down, so there's reasonable baseline security, and developers have to explicitly specify what sort of ingress/egress access they expect".

> ### Precedence model

Reviewer 11: ^^ Brief update from thoughts in the background while we were discussing this KEP today, even though we did not specifically discuss these proposals. Option (1) above is not explicit enough, and if it introduces a "default deny" then it begs the question how the policy applies to cluster-external traffic. We are not aiming to address cluster-external traffic in this KEP, and I don't think that it would be reasonable to define implicit behaviour on the policy impact for cluster-internal traffic only. I think it's pretty safe to discard Option (1) above.

> ### Precedence model

Author: since we moved on to a newer proposal including priorities we can resolve this comment as it no longer applies

#### Story 2: Funnel traffic through ingress/egress gateways

As a cluster admin, I want to ensure that all traffic coming into (going out of)
my cluster always goes through my ingress (egress) gateway.

It is common practice in enterprises to setup checkpoints in their clusters at
ingress/egress. These checkpoints usually perform advanced checks such as
firewalling, authentication, packet/connection logging, etc.
This is a big request for compliance reasons, and ClusterNetworkPolicy can ensure
that all the traffic is forced to go through ingress/egress gateways.
It is worth noting that the Cluster-scoped NetworkPolicy APIs will not redirect
traffic, rather it can ensure that no traffic is allowed in/out except traffic
via the gateways.

#### Story 3: Isolate multiple tenants in a cluster

As a cluster admin, I want to isolate all the tenants (modeled as Namespaces)
on my cluster from each other by default.

Many enterprises are creating shared Kubernetes clusters that are managed by a
centralized platform team. Each internal team that wants to run their workloads
gets assigned a Namespace on the shared clusters. Naturally, the platform team

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer: You are correct in that, to enforce such tenant isolation, we needed 'except' rules to counter the 'deny' rules. I have added a new 'authorize' action in the latest commit to capture this. 

This story, however, was set out to express a different intent: tenant isolation is _not_ enforced, but rather provided as a starting point (or guideline) for initial security postures in tenants. Tenants can actually completely overwrite this based on the need of its namespaces. Of course, they cannot overwrite `Allow`, `Deny` and `Authorize` rules which cluster admins listed out as guardrails (dns must be allowed, egress to some IPs must be denied etc.)

If the cluster admin wants to _enforce_ tenant isolation instead, then it's a complete different story. As a matter of fact, this becomes Story 4 below.

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer 12: I'm not sure it's covered by story 4. Story 4 talks about allowing cluster-wide which seems very different to the use case of enforcing strict boundaries across namespaces which cannot be broken by the tenants themselves. This is one of the most common use cases we see Cilium users using cluster-wide policies for.

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer: If you look at the updated sample specs for story 4, I have actually written out how a strict boundary across namespace can be enforced:
```yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        type: tenant  # assuming all tenant namespaces will be created with this label
  ingress:
    - action: Authorize
      from:
      - namespaces:
          self: true
    - action: Deny
      from:
      - namespaces:
          selector: {}
```

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer 5: I still think that this story doesn't truly justify a DNP, it only justifies existence because CNPs in the current proposal don't  enable a default-deny behavior.

If we aligned CNP to act more like existing network policies in that they apply a default-deny to selected pods, then this story no longer needs a new resource and can instead be implemented entirely with ClusterNetworkPolciy and the existing NetworkPolicy API.

e.g.,

- A single CNP that selects all pods, allows nothing (enabling default deny behavior).
- A NP per-namespace that selects all pods, and allows to/from all pods within the namespace.

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer 5: ^ I think that has the bonus win of making CNP and NP behave more intuitively. The fact that one has a default deny and the other doesn't feels like a real area for confusion.

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer: ```
- action: Deny 
  from: 
   - namespaces: 
        notSame: 
        - key: kubernetes.io/metadata.name
```
this sounds like neat alternative that we have not explored to me. I'll put more thought into it.

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer 4: Are there any use cases for "same"/"not-same" with CNPs?

It seems like the set of things people want to do with "allow/deny X by default" rules ("DNPs") doesn't really have a _huge_ overlap with the set of things people want to do with "forcibly allow/deny X" rules ("CNPs"), so maybe they shouldn't actually be so similar?

Like, there is little point in creating a DNP that says "all egress is blocked by default, except to CoreDNS", because as soon as the user creates their own NetworkPolicy to allow the egress they need, they have to remember to re-allow CoreDNS again, because their NP would override the DNP. It would be much simpler for everyone if the admin just created a DNP saying "all egress is blocked by default, period" and a CNP saying "egress to CoreDNS is allowed, always".

So, if CNPs don't need "same-namespace" and DNPs don't need... lots of things, then maybe DNP shouldn't be like CNP, but should be more tightly focused on just establishing the baseline isolation/tenancy model, using a set of fields/selectors that make sense for the baseline isolation/tenancy use cases.

> will want to make sure that, by default, all intra-namespace traffic is allowed

Reviewer 6: I think same/notSame in CNP enables multi-namespace tenancy in a way nothing else does.  For hard tenancy, you really don't want it to be user-overrideable.

IOW - a user in a tenant=coke namespace is allowed to talk to other tenant=coke namespace OR to tenant=shared; but they still have to use network policy to do so.  And they can talk to DNS

CNP = [ deny * except [ notSame(tenant), tenant=shared ] ; allow DNS ]
DNP = allow nothing
NP = <explicit>

The CNP expression is what worries me.  I don't want to "allow same(tenant)" because that is a unilateral allow (including things that user NP doesn't want).  I want to deny everything except same tenant and system stuff, both of which are "allowable" but not "allowed".  Can we express that more succinctly?

something like:

CNP = [ empower same(tenant); empower tenant=system; allow DNS ]

that seems more expressive - empowered != allowed

The same Pods, on the other hand, can appear as an AppliedTo, or an ingress/egress
peer in any ClusterNetworkPolicy. This makes looking up policies that affect a
particular Pod more challenging than when there's only NetworkPolicy resources.

In addition, in an extreme case where a ClusterNetworkPolicy `Deny` rule,
ClusterNetworkPolicy `Allow` rule, NetworkPolicy rule and DefaultNetworkPolicy
rule applies to an overlapping set of Pods, users will need to refer to the
precedence model mentioned in the [previous section](precedence-model) to
determine which rule would take effect. As shown in that section, figuring out
how stacked policies affect traffic between workloads might not be very straightfoward.

To mitigate this risk and improve UX, a tool which reversely looks up affecting
policies for a given Pod and prints out relative precedence of those rules
can be quite useful. The [cyclonus](https://github.com/mattfenwick/cyclonus) project
for example, could be extended to support ClusterNetworkPolicy and DefaultNetworkPolicy.
This is an orthogonal effort and will not be addressed by this KEP in particular.

> 

Reviewer 3: Here a question that came to my mind: we've discussed in past about CNIs providing some useful information about a NetworkPolicy on the status field. @danwinship proposed something here: https://github.com/kubernetes/enhancements/pull/2137

So, by our discussions I can't remember if the status was proposed/designed, but I guess that it is important that we try to predict this on the ClusterNetPol proposal.

In my mind, as a cluster admin, it would be good to have some feedback from the CNI controller at least if and how a NetworkPolicy has been parsed, as an example if it was fully implemented, if it got some parsing issue, etc.

The user stories provided [here](https://github.com/kubernetes/enhancements/pull/2137/files#diff-80527556c28fa329efb6e422217b3434c6b3d28efe9c840ca4f0905138a15aa7R179-R198) IMO are the most important to be taken care of a status field, and maybe even helping to report some misconfiguration as writen in https://github.com/kubernetes/enhancements/pull/2137/files#diff-80527556c28fa329efb6e422217b3434c6b3d28efe9c840ca4f0905138a15aa7R473

Reviewer 3: Just to clarify, as a user I don't expect metrics on the status field (time to apply, if this was applied in all nodes, etc) but a feedback from the Controller if the object was right, if all the selectors could be parsed, if there's some misconfiguration.

Author: i was actually hoping to see @danwinship proposal evolve and inherit the status part for CNPs :D it will be useful for users i agree

Author: updated KEP to include status

Cluster Administrator. It is not intended to solve:
1. Logging / error reporting for Network Policy
2. Kubernetes Node policies
3. New policy selectors (services, service accounts, etc.)
4. Tools/CLI to debug/explain NetworkPolicies, cluster-scoped policies and their
   impact on workloads.

## Proposal

In order to achieve the two primary broad use cases for a cluster admin to
secure K8s clusters, we propose to introduce the following two new resources
under `networking.k8s.io` API group:
- ClusterNetworkPolicy
- ClusterDefaultNetworkPolicy

### ClusterNetworkPolicy resource

A ClusterNetworkPolicy resource will help the administrators set strict
security rules for the cluster, i.e. a developer CANNOT override these rules
by creating NetworkPolicies that applies to the same workloads as the
ClusterNetworkPolicy does.

Unlike the NetworkPolicy resource in which each rule represents an allowed
traffic, ClusterNetworkPolicy will enable administrators to set `Authorize`,
`Deny` or `Allow` as the action of each rule. ClusterNetworkPolicy rules should
be read as-is, i.e. there will not be any implicit isolation effects for the Pods

> selected by the ClusterNetworkPolicy, as opposed to what NetworkPolicy rules imply.

Reviewer: See https://github.com/kubernetes/enhancements/pull/2522#discussion_r611962096

> In terms of precedence, the aggregated `Authorize` rules (all ClusterNetworkPolicy

Reviewer: Sure. I will follow up and change it to `AlwaysAllow`.

> In terms of precedence, the aggregated `Authorize` rules (all ClusterNetworkPolicy

Reviewer 5: I'm still... uneasy about overloading a single "action" field to include a concept of precedence as well as the pure action.

I guess let's see what the broader SIG / API reviewers have to say about this.

> In terms of precedence, the aggregated `Authorize` rules (all ClusterNetworkPolicy

Reviewer 6: Would it make more sense to simply go to `int` priority values so "exceptions" were less exceptional?

> In terms of precedence, the aggregated `Authorize` rules (all ClusterNetworkPolicy

Reviewer 9: I second revisiting int priority values so that the order of precedence is absolutely clear without various combinations of ambiguous keywords like Authorize, Except, AlwaysAllow and so on

> In terms of precedence, the aggregated `Authorize` rules (all ClusterNetworkPolicy

Reviewer 6: Recap of sig-net discussion:

I'm not hard-against any particular model.  What I want to balance is:

* simplicity (less is more)
* expressiveness (can I make it do what I need)
* familiarity (is it like something else I know)

If numeric priorities win big points on familiarity, then I am willing to overlook a bit more complexity. 

The Allow/Deny model with `exceptions` feels simple to me, but I ACK that maybe just me.

  This can be addressed by introducing a field, say `name`, per `ClusterNetworkPolicy`
  and `ClusterDefaultNetworkPolicy` ingress/egress rule.
- **Node Selector**: Cluster administrators and developers want to write
  policies that apply to cluster nodes or host network pods. This can be
  addressed by introducing nodeSelector field under `appliedTo` field of the
  `ClusterNetworkPolicy` and `ClusterDefaultNetworkPolicy` spec.
  `ClusterDefaultNetworkPolicy` is a better candidate compared to K8s `NetworkPolicy`
  for introducing this field as nodes are cluster level resources.

## Design Details

### ClusterNetworkPolicy API Design
The following new `ClusterNetworkPolicy` API will be added to the `networking.k8s.io` API group:

```golang
type ClusterNetworkPolicy struct {
	metav1.TypeMeta
	metav1.ObjectMeta

	Spec ClusterNetworkPolicySpec
}

type ClusterNetworkPolicySpec struct {
	// No implicit isolation of AppliedTo Pods.
	AppliedTo    AppliedTo
	Ingress      []ClusterNetworkPolicyIngressRule
	Egress       []ClusterNetworkPolicyEgressRule
}

type ClusterNetworkPolicyIngress/EgressRule struct {
	Action       RuleAction

> 	Ports        []networkingv1.NetworkPolicyPort

Reviewer 2: For the purpose of future extensibility this needs to be flipped.... Empty selectors should mean do Nothing

	From/To      []ClusterNetworkPolicyPeer
}

type ClusterNetworkPolicyPeer struct {
	PodSelector  *metav1.LabelSelector
	// required if a PodSelector is specified
	Namespaces   *Namespaces
	IPBlock      *networkingv1.IPBlock
}

const (
  RuleActionAuthorize RuleAction = "Authorize"
	RuleActionDeny      RuleAction = "Deny"
	RuleActionAllow     RuleAction = "Allow"
)
```

For the ClusterNetworkPolicy ingress/egress rule, the `Action` field dictates whether
traffic should be allowed or denied or always allowed from/to the ClusterNetworkPolicyPeer.
This will be a required field.

### ClusterDefaultNetworkPolicy API Design
The following new `ClusterDefaultNetworkPolicy` API will be added to the `networking.k8s.io` API group:

```golang
type ClusterDefaultNetworkPolicy struct {
	metav1.TypeMeta
	metav1.ObjectMeta

	Spec ClusterDefaultNetworkPolicySpec
}

type ClusterDefaultNetworkPolicySpec struct {
	// Implicit isolation of AppliedTo Pods.
	AppliedTo   AppliedTo
	Ingress     []ClusterDefaultNetworkPolicyIngressRule
	Egress      []ClusterDefaultNetworkPolicyEgressRule
}

type ClusterDefaultNetworkPolicyIngress/EgressRule struct {
	Ports            []networkingv1.NetworkPolicyPort

> 	OnlyFrom/OnlyTo  []ClusterDefaultNetworkPolicyPeer

Reviewer 5: > We chose this naming to better hint policy writers about the isolation effect of ClusterDefaultNetworkPolicy on the workloads it applies to.

To me, this feels like an unnecessary divergence from the existing NP API that will just trip up policy writers by forcing them to remember syntactical differences between ClusterDefaultNetworkPolicy and NetworkPolicy that are not consequential to the way the two resources behave.

Reviewer: Okay I will change this for the sake of consistency

Author: removed text

}

type ClusterDefaultNetworkPolicyPeer struct {

> 	PodSelector  *metav1.LabelSelector

Author: added a note regarding the motivation to remain somewhat familiar to NetPol

	// required if a PodSelector is specified
	Namespaces   *Namespaces
	IPBlock      *networkingv1.IPBlock
}
```

Most structs above are very similar to NetworkPolicy and quite self-explanatory.
One detail to notice is that in the ClusterDefaultNetworkPolicy Ingress/Egress rule spec,
the peers are created in a field named `OnlyFrom`/`OnlyTo`, as opposed to `To`/`From`
in ClusterNetworkPolicy. We chose this naming to better hint policy writers about
the isolation effect of ClusterDefaultNetworkPolicy on the workloads it applies to.

### Shared API Design
The following structs will be added to the `networking.k8s.io` API group and
shared between `ClusterNetworkPolicy` and `ClusterDefaultNetworkPolicy`:

```golang
type AppliedTo struct {
	// required if a PodSelector is specified
	NamespaceSelector   *metav1.LabelSelector
	// optional
	PodSelector         *metav1.LabelSelector
}

type Namespaces struct {
	// Self and Selector are mutually-exclusive

> 	Self       bool

Reviewer 5: Generally nice to avoid using bool fields in favor of enums. It makes it easier to extend the field's behavior in the future.

e.g.,

```go
type Namespaces struct {
  Match NamespaceMatchType
}

type NamespaceMatchType string

const (
  NamespaceMatchSelf NamespaceMatchType = "Self"
  
  // Contrived example of something that might be added in the future.
  NamespaceMatchNotSelf NamespaceMatchtype = "NotSelf"
)
```

Reviewer: We considered this and decided to go with the `bool` field because as of now there aren't other `NamespaceMatchtype` that makes sense. For the `NamespaceMatchNotSelf`, are you suggesting its meaning as "select from all namespaces except from self Namespaces"? That will be an interesting semanitcs. 
Also as @tnqn pointed out in https://github.com/kubernetes/enhancements/pull/2522#discussion_r614727344 how namespace matching and namespaceSelector co-exist with each other is another thing to consider

Reviewer: > Self, Selector, SameLabels, DifferentLabels - I can name 4 off the top of my head :)

Well the problem is the other `NamespaceMatchtypes` you suggested here, user will need to provide a kv pair to argument that `NamespaceMatchtype`, correct? Whereas for `self`, this isn't necessary and ideally this should be as simple as `Namespaces: self`.
This is what we were stuggling with before deciding to go with the bool field: how those special cases/keywords can co-exist with the current MatchLabels/MatchExpression structs.

Reviewer 6: Yes, some of those would take params, and that is OK.  e.g.

```
namespaces:
  scope: Self
```

vs

```
namespaces:
  scope: Selector
  selector:
  - foo: bar
```

vs

```
namespaces:
  scope: SameLabels
  labels:
  - mycompany.com/tenant
```

	Selector   *metav1.LabelSelector
}
```

#### AppliedTo

> The `AppliedTo` field replaces `PodSelector` in NetworkPolicy spec, as means to specify

Reviewer 5: > The `AppliedTo` field replaces `PodSelector` in NetworkPolicy spec,

Is this suggesting a change to the existing NetworkPolicy spec? The wording sounds like it, but my understanding is that we're not intending to touch the existing API as part of this.

Reviewer: Will re-word

    # gateway namespace will be sufficient
    namespaceSelector: {}
  ingress:
    - action: Deny
      from:
      - ipBlock:
          cidr: 192.0.2.0/24  # banned addresses
```

#### Story 2: Funnel traffic through ingress/egress gateways
As a cluster admin, I want to ensure that all traffic coming into (going out of)
my cluster always goes through my ingress (egress) gateway.

```yaml
apiVersion: networking.k8s.io/v1alpha1
kind: ClusterNetworkPolicy
metadata:
  name: ingress-egress-gateway
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        type: tenant  # assuming all tenant namespaces will be created with this label
  ingress:
    - action: Authorize
      from:
      - namespaces:
          selector:
            matchLabels:
              kubernetes.io/metadata.name: dmz # ingress gateway
    - action: Deny
      from:
      - ipBlock:
          cidr: 0.0.0.0/0
  egress:
    - action: Authorize
      to:
      - namespaces:
          selector:
            matchLabels:
              kubernetes.io/metadata.name: istio-egress  # egress gateway
    - action: Deny
      to:
      - ipBlock:
          cidr: 0.0.0.0/0
```

> __Note:__ The above policy is very restrictive, i.e. it rejects ingress/egress

Reviewer: > I don't think we can use such a model as an example - it's not a great solution to the problem presented :)

Because it is too restrictive to start with?

If a Pod is selected by both a DefaultNetworkPolicy and a NetworkPolicy, then the DefaultNetworkPolicy's
effect on that Pod becomes obsolete. The traffic allowed will be solely determined by the NetworkPolicy.

(TODO: Add a diagram to explain the precedence?)

> Together, a ClusterNetworkPolicy and a DefaultNetworkPolicy resource can help satisfy all the

Author: sorry @Dyanngg this got removed. could you please re-add :)

                            # the control of intra-namespace traffic and from system namespaces like kube-system.
spec:
  appliedTo:
    namespaceSelector:
      matchLabels:
        type: tenant      # assuming all tenant namespaces will be created with this label
  ingress:
    - from:
      - namespaces:
          scope: Self
        action: Empower   # add exception to deny rule for intra-namespace traffic
      - namespaceSelector:
          matchLabels:
            app: system
        action: Empower   # add exception to deny rule for system namespaces like kube-system
      - action: Deny      # deny traffic from all namespaces except for exceptions provided by Empower rules
        namespaceSelector: {}
```

In this policy, tenant isolation (Namespace being the boundary) is strictly enforced.
Tenants can however allow or deny intra-namespace traffic and traffic from "kube-system"
Namespace depending on their needs. They cannot, however, overwrite Allow and Deny
rules which cluster admins listed out as guardrails (dns must be allowed, egress
to some IPs must be denied etc.)

> #### Sample Spec for Story 4: Enforce network/security best practices

Reviewer: I wonder if we should revise the name of this story now. Instead of `Enforce network/security best practices`, it feels more like something in the lines of `Set zero-trust default security posture for tenants`

Author: done

### Test Plan

- Add e2e tests for ClusterNetworkPolicy resource

>   - Ensure `Empower`rules are excepted

Reviewer: Ensure `Empower` rules override `Deny` rules and provide exceptions to the `Deny` rules?

Author: done
